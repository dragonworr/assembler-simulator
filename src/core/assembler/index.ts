import { produce } from 'immer'
import { tokenize } from './tokenizer'
import type { Statement } from './parser'
import { OperandType, parse } from './parser'
import {
  DuplicateLabelError,
  EndOfMemoryError,
  LabelNotExistError,
  JumpDistanceError
} from '../exceptions'
import { Mnemonic } from '../constants'
import { exp } from '../../common/utils'

export * from './tokenizer'
export * from './parser'

type LabelToAddressMap = Record<string, number>

const getLabelToAddressMap = (statements: Statement[]): LabelToAddressMap => {
  const [, labelToAddressMap] = statements.reduce<[number, LabelToAddressMap]>(
    ([address, labelToAddressMap], statement, index) => {
      const { label, instruction, operands, machineCodes } = statement
      const firstOperand = operands[0]
      return [
        instruction.token.value === Mnemonic.ORG
          ? (firstOperand.value as number)
          : exp<number>(() => {
              const nextAddress =
                address +
                machineCodes.length +
                (firstOperand !== undefined && firstOperand.type === OperandType.Label ? 1 : 0)
              if (nextAddress > 0xff && index !== statements.length - 1) {
                throw new EndOfMemoryError(statement)
              }
              return nextAddress
            }),
        label === null
          ? labelToAddressMap
          : produce(labelToAddressMap, draft => {
              if (draft[label.identifier] !== undefined) {
                throw new DuplicateLabelError(label)
              }
              draft[label.identifier] = address
            })
      ]
    },
    [0, {}]
  )
  return labelToAddressMap
}

export type AddressToMachineCodeMap = Record<number, number>
export type AddressToStatementMap = Record<number, Statement>

type AssembleResult = [AddressToMachineCodeMap, AddressToStatementMap]

export const assemble = (input: string): AssembleResult => {
  const statements = parse(tokenize(input))
  const labelToAddressMap = getLabelToAddressMap(statements)
  const [, addressToMachineCodeMap, addressToStatementMap] = statements.reduce<
    [number, ...AssembleResult]
  >(
    ([address, addressToMachineCodeMap, addressToStatementMap], statement) => {
      const { instruction, operands, machineCodes } = statement
      const firstOperand = operands[0]
      if (instruction.token.value === Mnemonic.ORG) {
        return [firstOperand.value as number, addressToMachineCodeMap, addressToStatementMap]
      }
      if (firstOperand !== undefined && firstOperand.type === OperandType.Label) {
        const labelAddress = labelToAddressMap[firstOperand.token.value]
        if (labelAddress === undefined) {
          throw new LabelNotExistError(firstOperand.token)
        }
        const distance = labelAddress - address
        if (distance < -128 || distance > 127) {
          throw new JumpDistanceError(firstOperand.token)
        }
        const unsignedDistance = distance < 0 ? 0x100 + distance : distance
        machineCodes.push(unsignedDistance)
      }
      return [
        address + machineCodes.length,
        produce(addressToMachineCodeMap, draft => {
          machineCodes.forEach((machineCode, index) => {
            draft[address + index] = machineCode
          })
        }),
        produce(addressToStatementMap, draft => {
          draft[address] = statement
        })
      ]
    },
    [0, {}, {}]
  )
  return [addressToMachineCodeMap, addressToStatementMap]
}
