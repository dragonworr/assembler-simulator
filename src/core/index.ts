export { assemble } from './assembler'
export * from './memory'
export * from './cpu'
