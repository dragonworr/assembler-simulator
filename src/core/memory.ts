import type { AddressToMachineCodeMap } from './assembler'

export const initMemory = (): number[] =>
  [...Array(0x100)].map((_, address) => (address < 0xc0 ? 0 : 0x20))

export const initMemoryFrom = (map: AddressToMachineCodeMap): number[] => {
  const memory = initMemory()
  map.forEach((machineCode, address) => {
    memory[address] = machineCode
  })
  return memory
}
