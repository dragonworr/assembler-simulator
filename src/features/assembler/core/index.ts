import { produce } from 'immer'
import { tokenize } from './tokenizer'
import { OperandType, Statement, parse } from './parser'
import {
  DuplicateLabelError,
  AssembleEndOfMemoryError,
  LabelNotExistError,
  JumpDistanceError
} from '../../../common/exceptions'
import { Mnemonic } from '../../../common/constants'
import { call } from '../../../common/utils'

export * from './types'
export * from './tokenizer'
export * from './parser'

type LabelToAddressMap = Record<string, number>

const getLabelToAddressMap = (statements: Statement[]): LabelToAddressMap => {
  const [, labelToAddressMap] = statements.reduce<[number, LabelToAddressMap]>(
    ([address, labelToAddressMap], statement, index) => {
      const { label, instruction, operands, machineCodes } = statement
      const firstOperand = operands[0]
      return [
        instruction.mnemonic === Mnemonic.ORG
          ? (firstOperand.value as number)
          : call((): number => {
              const nextAddress =
                address +
                machineCodes.length +
                (firstOperand !== undefined && firstOperand.type === OperandType.Label ? 1 : 0)
              if (nextAddress > 0xff && index !== statements.length - 1) {
                throw new AssembleEndOfMemoryError(statement)
              }
              return nextAddress
            }),
        label === null
          ? labelToAddressMap
          : produce(labelToAddressMap, draft => {
              if (draft[label.identifier] !== undefined) {
                throw new DuplicateLabelError(label)
              }
              draft[label.identifier] = address
            })
      ]
    },
    [0, {}]
  )
  return labelToAddressMap
}

export type AddressToMachineCodeMap = Record<number, number>
export type AddressToStatementMap = Record<number, Statement>

type AssembleResult = [AddressToMachineCodeMap, AddressToStatementMap]

export const assemble = (input: string): AssembleResult => {
  const statements = parse(tokenize(input))
  const labelToAddressMap = getLabelToAddressMap(statements)
  const [, addressToMachineCodeMap, addressToStatementMap] = statements.reduce<
    [number, ...AssembleResult]
  >(
    ([address, addressToMachineCodeMap, addressToStatementMap], statement) => {
      const { instruction, operands, machineCodes } = statement
      const firstOperand = operands[0]
      if (instruction.mnemonic === Mnemonic.ORG) {
        return [firstOperand.value as number, addressToMachineCodeMap, addressToStatementMap]
      }
      if (firstOperand !== undefined && firstOperand.type === OperandType.Label) {
        const labelAddress = labelToAddressMap[firstOperand.rawValue]
        if (labelAddress === undefined) {
          throw new LabelNotExistError(firstOperand)
        }
        const distance = labelAddress - address
        if (distance < -128 || distance > 127) {
          throw new JumpDistanceError(firstOperand)
        }
        const unsignedDistance = distance < 0 ? 0x100 + distance : distance
        machineCodes.push(unsignedDistance)
      }
      return [
        address + machineCodes.length,
        produce(addressToMachineCodeMap, draft => {
          machineCodes.forEach((machineCode, index) => {
            draft[address + index] = machineCode
          })
        }),
        produce(addressToStatementMap, draft => {
          draft[address] = statement
        })
      ]
    },
    [0, {}, {}]
  )
  return [addressToMachineCodeMap, addressToStatementMap]
}
