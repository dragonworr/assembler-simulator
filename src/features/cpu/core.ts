import { produce } from 'immer'
import {
  InvalidRegisterError,
  RunBeyondEndOfMemory,
  // StackOverflowError,
  // StackUnderflowError,
  DivideByZeroError
} from '../../common/exceptions'
import { Opcode, Register } from '../../common/constants'
import { Head, unsign8bit, exp } from '../../common/utils'

type GPR = [AL: number, BL: number, CL: number, DL: number]

const MAX_SP = 0xbf

enum Flag {
  Zero,
  Overflow,
  Sign,
  Interrupt
}

type SR = [zero: boolean, overflow: boolean, sign: boolean, interrupt: boolean]

interface CPU {
  gpr: GPR
  ip: number
  sp: number
  sr: SR
  isHalted: boolean
}

export const init = (): CPU => {
  return {
    gpr: [0, 0, 0, 0],
    ip: 0,
    sp: MAX_SP,
    sr: [false, false, false, false],
    isHalted: false
  }
}

const checkGPR = (register: number): Register => {
  if (register < 0 || register > 3) {
    throw new InvalidRegisterError(register)
  }
  return register
}

const checkIP = (address: number): number => {
  if (address > 0xff) {
    throw new RunBeyondEndOfMemory()
  }
  return address
}

// const checkSP = (address: number): number => {
//   if (address < 0) {
//     throw new StackOverflowError()
//   }
//   if (address > MAX_SP) {
//     throw new StackUnderflowError()
//   }
//   return address
// }

// const getFlagsValue = (sr: SR): number =>
//   sr.reduce((result, isSet, flag) => (isSet ? result + 0b10 ** flag : result), 0)

const checkOperationResult = (
  result: number,
  previousValue: number
): [result: number, flags: Partial<SR>] => {
  const finalResult = result > 0xff ? result % 0x100 : unsign8bit(result)
  const flags: Head<SR> = [false, false, false]
  if (finalResult === 0) {
    flags[Flag.Zero] = true
  } else if (finalResult >= 0x80) {
    flags[Flag.Sign] = true
  }
  if (
    /* addition */ (previousValue < 0x80 && result >= 0x80) ||
    /* subtraction */ (previousValue >= 0x80 && result < 0x80)
  ) {
    flags[Flag.Overflow] = true
  }
  return [finalResult, flags]
}

export const step = (__cpu: CPU, __memory: number[]): [cpu: CPU, memory: number[]] =>
  produce([__cpu, __memory], ([cpu, memory]) => {
    const getGPR = (register: Register): number => cpu.gpr[register]
    const setGPR = (register: Register, value: number): void => {
      cpu.gpr[register] = value
    }

    const setIP = (address: number): void => {
      cpu.ip = address
    }
    const incIP = (): void => {
      setIP(checkIP(cpu.ip + 1))
    }

    const setSR = (flags: Partial<SR>): void => {
      Object.assign(cpu.sr, flags)
    }

    const loadFromMemory = (address: number): number => {
      return memory[address]
    }
    // const storeToMemory = (address: number, machineCode: number): void => {
    //   memory[address] = machineCode
    // }

    /**
     * @modifies {@link cpu.ip}
     */
    const getNextOpcode = (): number => {
      incIP()
      return loadFromMemory(cpu.ip)
    }

    const opcode = loadFromMemory(cpu.ip)

    switch (opcode) {
      case Opcode.END:
        cpu.isHalted = true
        break
      case Opcode.ADD_REG_TO_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) + getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.SUB_REG_FROM_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) - getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.MUL_REG_BY_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) * getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.DIV_REG_BY_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          exp<number>(() => {
            const divisor = getGPR(srcReg)
            if (divisor === 0) {
              throw new DivideByZeroError()
            }
            return Math.floor(getGPR(destReg) / divisor)
          }),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.INC_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(getGPR(destReg) + 1, getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.DEC_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(getGPR(destReg) - 1, getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.MOD_REG_BY_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          exp<number>(() => {
            const divisor = getGPR(srcReg)
            if (divisor === 0) {
              throw new DivideByZeroError()
            }
            return getGPR(destReg) % divisor
          }),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.AND_REG_WITH_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) & getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.OR_REG_WITH_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) | getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.XOR_REG_WITH_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) ^ getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.NOT_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(~getGPR(destReg), getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.ROL_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          exp<number>(() => {
            const value = getGPR(destReg)
            const MSB = Math.floor(value / 0x80)
            return (value << 1) + MSB
          }),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.ROR_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          exp<number>(() => {
            const value = getGPR(destReg)
            const LSB = value % 2
            return LSB * 0x80 + (value >> 1)
          }),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.SHL_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(getGPR(destReg) << 1, getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.SHR_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(getGPR(destReg) >> 1, getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
    }
  })
