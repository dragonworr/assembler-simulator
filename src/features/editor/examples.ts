const COMMENT_DIVIDER = '; --------------------------------------'

const applyTemplate = (title: string, body: string, footer = ''): string => `${COMMENT_DIVIDER}
;\t${title}
${COMMENT_DIVIDER}
${body}
${COMMENT_DIVIDER}
\tEND
${COMMENT_DIVIDER}
${footer}`

export const NEW_FILE_TEMPLATE = applyTemplate('New File', '\t')

const FOOTER_MESSAGE = 'For more examples, select File > Open Example.'

interface Example {
  title: string
  content: string
}

const createExample = (title: string, body: string): Example => {
  return {
    title,
    content: applyTemplate(title, body, `\n${FOOTER_MESSAGE}\n`)
  }
}

export const examples: readonly Example[] = [
  createExample(
    'Keyboard Input',
    `\tMOV BL, C0\t\t; Start address of VDU

Loop:
\tIN  00\t\t\t; Wait for keyboard input
\tCMP AL, 0D\t\t; Check if the key was Enter (Carriage Return)
\tJZ  Done\t\t; Yes => Jump to Done

\tMOV [BL], AL\t; Output to VDU
\tINC BL\t\t\t; Increase the address of VDU by one
\tJNZ Loop\t\t; Jump back if there is still an available address

Done:`
  ),
  createExample(
    'Visual Display Unit',
    `\tJMP Start

\tDB "Hello World!"
\tDB 00

Start:
\tMOV AL, C0\t\t; Start address of VDU
\tMOV BL, 02\t\t; Address of the first character in the string
\tMOV CL, [BL]\t; Store the first character in CL

Loop:
\tMOV [AL], CL\t; Output to VDU
\tINC AL\t\t\t; Increase the address of VDU by one
\tINC BL\t\t\t; Increase the address of character by one
\tMOV CL, [BL]\t; Store the next character in CL
\tCMP CL, 00\t\t; Check if the next character is zero (NULL)
\tJNZ Loop\t\t; NO => Jump back to Loop`
  ),
  createExample(
    'Traffic Lights',
    `Start:
\tMOV AL, 80\t; 1000 0000

Loop:
\tOUT 01\t\t; Output to Traffic Lights
\tDIV AL, 02\t; Move the bit to right by one
\tCMP AL, 00\t; Check if AL is zero
\tJNZ Loop\t; No => Jump back to Loop
\tJMP Start`
  ),
  createExample(
    'Seven-Segment Display',
    `\tJMP Start

\tDB FA\t\t\t; 0 - 1111 1010
\tDB 60\t\t\t; 1 - 0110 0000
\tDB B6\t\t\t; 2 - 1011 0110
\tDB 9E\t\t\t; 3 - 1001 1110
\tDB 4E\t\t\t; 4 - 0100 1110
\tDB DC\t\t\t; 5 - 1101 1100
\tDB FC\t\t\t; 6 - 1111 1100
\tDB 8A\t\t\t; 7 - 1000 1010
\tDB FE\t\t\t; 8 - 1111 1110
\tDB DE\t\t\t; 9 - 1101 1110
\tDB 00

Start:
\tMOV BL, 02\t\t; Start address of the data table
\tMOV AL, [BL]\t; Store the first number in AL

Loop:
\tOUT 02\t\t\t; Output to Seven-segment Display
\tINC AL\t\t\t; Set LSB of the number to one
\tNOP NOP NOP\t\t; Wait for three cycles
\tOUT 02\t\t\t; Output to Seven-segment Display
\tINC BL\t\t\t; Increase the address of the data table by one
\tMOV AL, [BL]\t; Store the next number in AL
\tCMP AL, 00\t\t; Check if the next number is zero (NULL)
\tJNZ Loop\t\t; No => Jump back to Loop
\tJMP Start`
  )
]
