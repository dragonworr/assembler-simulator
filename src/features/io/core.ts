export const SKIP = -1

export const MAX_PORT = 0x0f

export enum InputPort {
  SimulatedKeyboard = 0,
  Thermostat = 3,
  Keyboard = 7,
  NumericKeypad = 8
}

export type InputData =
  | {
      content: number
      port: InputPort
    }
  | {
      content: null
      port: null
    }

export interface InputSignals {
  data: InputData
  interrupt: boolean
  // TODO: interruptVectorAddress
}

export const initialInputSignals: InputSignals = {
  data: {
    content: null,
    port: null
  },
  interrupt: false
}

export enum OutputPort {
  TrafficLights = 1,
  SevenSegmentDisplay = 2,
  Heater = 3,
  SnakeInMaze = 4,
  StepperMotor = 5,
  Lift = 6,
  Keyboard = 7,
  NumericKeypad = 8
}

interface OutputData {
  content: number
  port: OutputPort
}

export type OutputSignals = Partial<{
  halted: true
  requiredInputPort: InputPort
  data: OutputData
  closeWindows: true
}>

export interface Signals {
  input: InputSignals
  output: OutputSignals
}
