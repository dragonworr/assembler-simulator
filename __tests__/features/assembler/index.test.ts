import { AssemblerError, assemble } from '@/features/assembler/core'
import { examples } from '@/features/editor/examples'
import { initDataFrom } from '@/features/memory/core'
import { memorySerializer } from '../../snapshotSerializers'

expect.addSnapshotSerializer(memorySerializer)

describe('assembler', () => {
  examples.forEach(({ title, content }) => {
    it(`should assemble example ${title}`, () => {
      const [addressToMachineCodeMap] = assemble(content)
      expect(initDataFrom(addressToMachineCodeMap)).toMatchSnapshot()
    })
  })

  it('should throw instance of AssemblerError', () => {
    try {
      assemble('foo')
    } catch (err) {
      expect(err).toBeInstanceOf(AssemblerError)
    }
    expect.assertions(1)
  })

  it('should throw DuplicateLabelError', () => {
    expect(() => {
      assemble(`
start: inc al
start: dec bl
end
`)
    }).toThrowError("Duplicate label 'START'")
  })

  it('should throw EndOfMemoryError', () => {
    expect(() => {
      assemble(`
org ff
inc al
end
`)
    }).toThrowError('Can not generate code beyond the end of RAM')
  })

  it('should throw LabelNotExistError', () => {
    expect(() => {
      assemble('jmp start end')
    }).toThrowError("Label 'start' does not exist")
  })

  it('should throw JumpDistanceError', () => {
    expect(() => {
      assemble(`
start:
inc al
org fd
jmp start
end
`)
    }).toThrowError('Jump distance should be between -128 and 127')
  })
})
