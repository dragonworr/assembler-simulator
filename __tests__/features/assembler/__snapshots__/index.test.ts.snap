// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assembler should assemble example Seven-Segment Display 1`] = `
[
  C0, 0D, FA, 60, B6, 9E, 4E, DC, FC, 8A, FE, DE, 00, D0, 01, 02,
  D3, 00, 01, F1, 02, A4, 00, FF, FF, FF, F1, 02, A4, 01, D3, 00,
  01, DB, 00, 00, C2, EF, C0, E7, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
]
`;

exports[`assembler should assemble example Traffic Lights 1`] = `
[
  D0, 00, 80, F1, 01, B3, 00, 02, DB, 00, 00, C2, F8, C0, F3, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
]
`;

exports[`assembler should assemble example Visual Display Unit 1`] = `
[
  C0, 0F, 48, 65, 6C, 6C, 6F, 20, 57, 6F, 72, 6C, 64, 21, 00, D0,
  00, C0, D0, 01, 02, D3, 02, 01, D4, 00, 02, A4, 00, A4, 01, D3,
  02, 01, DB, 02, 00, C2, F3, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
]
`;
